File: leetcode/java/20.ValidParenthesis.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> match = new HashMap<Character, Character>();
        match.put(')', '(');
        match.put('}', '{');
        match.put(']', '[');
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
                stack.push(s.charAt(i));
                continue;
            }
            if (stack.size() == 0 || match.get(s.charAt(i)) != stack.pop()) {
                return false;
            }
        }
        if (stack.size() == 0) {
            return true;
        }
        return false;
    }
}

File: leetcode/kotlin/20.ValidParenthesis.kt
import java.util.*
import kotlin.collections.HashMap

internal class Solution {
    fun isValid(s: String): Boolean {
        val match = HashMap<Char, Char>()
        match[')'] = '('
        match['}'] = '{'
        match[']'] = '['
        val stack = Stack<Char>()
        for (i in s.indices) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                stack.push(s[i])
                continue
            }
            if (stack.size == 0 || match[s[i]] !== stack.pop()) {
                return false
            }
        }
        return stack.size == 0
    }
}